File: cpp/362.design-hit-counter.cpp
class HitCounter {
    queue<pair<int, int>> hits;
    int current_sz = 0;
public:
    HitCounter() {
        
    }
    
    void hit(int timestamp) {
	    if(hits.empty()){
		    hits.push({timestamp, 1});
		    current_sz++;
		    return;
	    }
	    if(hits.back().first == timestamp){
		    hits.back().second++;
		    current_sz++;
		    return;
	    }
	    hits.push({timestamp, 1});
	    current_sz++;
    }
    
    int getHits(int timestamp) {
	while(!hits.empty() && timestamp-hits.front().first>=300 ){
		current_sz -= hits.front().second;
		hits.pop();
	}
	return current_sz;
    }
};

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter* obj = new HitCounter();
 * obj->hit(timestamp);
 * int param_2 = obj->getHits(timestamp);
 */
// 1 2 3   300 
//       4     300 301
//       3     4   3  

File: python/362.design-hit-counter.py
class HitCounter:
    def __init__(self):
        self.hits = collections.deque()
        self.count = 0

    def hit(self, timestamp: int) -> None:
        self.count+=1
        if len(self.hits) == 0:
            self.hits.append([timestamp, 1])
            return
        if self.hits[-1][0] == timestamp:
            self.hits[-1][1] +=1
            return
        self.hits.append([timestamp, 1])

    def getHits(self, timestamp: int) -> int:
        while len(self.hits) > 0 and timestamp-self.hits[0][0] >= 300:
            self.count -= self.hits[0][1]
            self.hits.popleft()
        return self.count

# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)
